File: app/src/main/java/com/example/phonebook/viewModels/MainActivityViewModel.java
package com.example.phonebook.viewModels;

import android.app.Application;

import android.widget.Toast;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.phonebook.model.data.Contact;
import com.example.phonebook.model.repos.ContactsRepoByRoom;
import com.example.phonebook.model.repos.ServiceRepoShPref;

import java.util.ArrayList;

import javax.inject.Inject;

public class MainActivityViewModel extends ViewModel {

    private ContactsRepoByRoom mContactRepo;
    private ServiceRepoShPref mServiceRepo;
    private Contact mContact;
    private Application mApplication;

    private ArrayList<Contact> mCurrentContactList;
    private MutableLiveData<ArrayList<Contact>> mContacts;

    @Inject
    public MainActivityViewModel(ContactsRepoByRoom contactsRepo, ServiceRepoShPref serviceRepo, Contact contact, Application application) {
        mContactRepo = contactsRepo;
        mServiceRepo = serviceRepo;
        mContact = contact;
        mApplication = application;
        mCurrentContactList = new ArrayList<>();
        mContacts = new MutableLiveData<>();
    }

    public Contact getContact() {
        return mContact;
    }

    public void setContact(Contact _contact) {
        this.mContact = _contact;
    }

    public MutableLiveData<ArrayList<Contact>> getCurrentContactList() {
        return mContacts;
    }

    public void init() {
        if(mContactRepo.getContactList(mServiceRepo.getCurrentUser()) != null)
            mCurrentContactList.addAll(mContactRepo.getContactList(mServiceRepo.getCurrentUser()));
        mContacts.postValue(mCurrentContactList);
    }

    public boolean addContact() {
        if(mContact.getContactName().length() == 0) {
            Toast mErrorNoName = Toast.makeText(mApplication, "Contact is empty", Toast.LENGTH_SHORT);
            mErrorNoName.show();
            return false;
        }

        Contact contact = new Contact();

        contact.setContactName(mContact.getContactName());
        contact.setContactPhone(mContact.getContactPhone());
        contact.setContactEmail(mContact.getContactEmail());
        contact.setContactAddress(mContact.getContactAddress());

        int index = mCurrentContactList.indexOf(mContact);

        if(index < 0)
            mCurrentContactList.add(contact);
        else
            mCurrentContactList.set(index, contact);

        mContacts.postValue(mCurrentContactList);
        mContact.clean();

        return true;
    }

    public void deleteContact() {
        mCurrentContactList.remove(mContact);
        mContacts.postValue(mCurrentContactList);
        mContact.clean();
    }

    public void signOut() {
        mServiceRepo.setSignedUp(false);
    }

    public void save() {
        mContactRepo.saveContactList(mServiceRepo.getCurrentUser(), mCurrentContactList);
    }
}

File: app/src/main/java/com/example/phonebook/viewModels/MainActivityViewModel.kt
package com.example.phonebook.viewModels

import android.app.Application
import android.util.Log

import android.widget.Toast

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel

import com.example.phonebook.model.data.Contact
import com.example.phonebook.model.repos.ContactsRepoByRoom
import com.example.phonebook.model.repos.ServiceRepoShPref

import java.util.ArrayList

import javax.inject.Inject

class MainActivityViewModel
@Inject constructor(
        private var mContactRepo: ContactsRepoByRoom,
        private var mServiceRepo: ServiceRepoShPref,
        var mContact: Contact,
        private var mApplication: Application): ViewModel() {

    var mCurrentContactList: ArrayList<Contact> = ArrayList()
    var mContacts: MutableLiveData<ArrayList<Contact>> = MutableLiveData()

    fun init() {
        mCurrentContactList.addAll(mContactRepo.getContactList(mServiceRepo.getCurrentUser() ?: "")?.asIterable() ?: ArrayList())
        mContacts.postValue(mCurrentContactList)
    }

    fun addContact(): Boolean {

        if(mContact.contactName.isEmpty()) {
            Toast.makeText(mApplication, "Contact is empty", Toast.LENGTH_SHORT).show()
            return false
        }

        val contact = Contact().apply {
            contactName = mContact.contactName
            contactPhone = mContact.contactPhone
            contactEmail = mContact.contactEmail
            contactAddress = mContact.contactAddress
        }

        val index = mCurrentContactList.indexOf(mContact)

        if(index < 0)
            mCurrentContactList.add(contact)
        else
            mCurrentContactList[index] = contact

        mContacts.postValue(mCurrentContactList)
        mContact.clean()
        return true
    }

    fun deleteContact() {
        mCurrentContactList.remove(mContact)
        mContacts.postValue(mCurrentContactList)
        mContact.clean()
    }

    fun signOut() = mServiceRepo.setSignedUp(false)

    fun save() = mContactRepo.saveContactList(mServiceRepo.getCurrentUser() ?: "", mCurrentContactList)
}
